relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+    ///
+    /// Except in testing, this should always be [`UnixTimestamp::now`].
+    pub start: UnixTimestamp,
+    /// Size of each individual bucket, inputs are truncated to this value.
+    pub bucket_interval: u32,
+    /// The amount of time slots to keep track off in the aggregator.
+    ///
+    /// The size of a time slot is defined by [`Self::bucket_interval`].
+    pub num_time_slots: u32,
+    /// The amount of partitions per time slot.
+    pub num_partitions: u32,
+    /// Delay how long a bucket should remain in the aggregator before being flushed.
+    ///
+    /// Ideally the delay is a multiple of [`Self::bucket_interval`].
+    pub delay: u32,
+    /// Maximum amount of bytes the aggregator can grow to.
+    pub max_total_bucket_bytes: Option<u64>,
+    /// Maximum amount of bytes the aggregator allows per project key.
+    pub max_project_key_bucket_bytes: Option<u64>,
+    /// The age in seconds of the oldest allowed bucket timestamp.
+    pub max_secs_in_past: Option<u64>,
+    /// The time in seconds that a timestamp may be in the future.
+    pub max_secs_in_future: Option<u64>,
+}
+
+/// A metrics aggregator.
+///
+/// The aggregator is unaware of current time and needs to be driven by periodic flushes using
+/// [`Self::flush_next`]. Each flush advances the internal clock by the configured
+/// [`Config::bucket_interval`].
+///
+/// The internal time is set on construction to [`Config::start`].
+///
+/// Use [`Self::next_flush`] to determine the time when to call [`Self::flush_next`].
+pub struct Inner {
+    /// Ring-buffer of aggregation slots.
+    ///
+    /// This is treated as a ring-buffer of a two dimensional array. The first dimension is a
+    /// "time slot" and the second dimension is the assigned partition.
+    ///
+    /// The total length of the ring-buffer is then determined by the amount of time slots times
+    /// the amount of partition. In other words, every time slot as [`Self::num_partitions`]
+    /// partitions.
+    ///
+    /// Layout:
+    /// Time slots: [            ][            ][            ]
+    /// Partitions:  [    ][    ]  [    ][    ]  [    ][    ]
+    ///
+    /// An item is assigned by first determining it's time slot and then assigning it a partition
+    /// based on selected [`Self::partition_by`] strategy.
+    ///
+    /// The first item in the buffer is tracked by [`Self::head`] which is at any time the
+    /// current partition since the beginning "zero". The beginning in the aggregator refers to the
+    /// beginning of the epoch. The [`Self::head`] is calculated with `f(x) = x / bucket_interval * num_partitions`.
+    ///
+    /// Flushing a partition advances the [`Self::head`] by a single value `+1`. Meaning
+    /// effectively time is advanced by `bucket_interval / num_partitions`.
+    slots: VecDeque<Slot>,
+    /// The amount of partitions per time slot.
+    num_partitions: u32,
+
+    /// Position of the first element in [`Self::slots`].
+    head: u64,
```

Comment:
I would encapsulate `slots` and `head` in a simple data structure that manages access. Something like 

```rust
slots.get(time_slot, partition)
```